### ISSUE TITLE
repeating websocket disconnect (code 1006) on W10 + chrome

### ISSUE THREAD

**Issue by brubbel** [django/channels#1407]
TLDR; When using websockets (Daphne websocket server), some clients seem to disconnect every 20 seconds with error code 1006.

Problem (Tested on 2 W10 pc's, 1 linux Mint, 1 Ubuntu and 1 W7 Virtualbox)
On the 2 W10 pc's using Chrome (e.g. 79.0.3945.130), it seems that websockets clients are terminated every 20 seconds. This is not the case with Chrome on Debian, Mint or W7.

Cause:
From Wireshark, it seems that the automatic ping_interval (setup in daphne/server), which sends a regular WebSocket ping (https://tools.ietf.org/html/rfc6455#section-5.5.2) to the client seems to drop the connection consistenly and immediately thereafter (https://tools.ietf.org/html/rfc6455#section-5.5.1) with code 1006.

Solution:
Do not use built-in rfc6455 websocket pings and use custom-designed pings over the websocket channel. Connections are stable now.

**Comment by qeternity** [comment]
This is really strange, we are experiencing the same thing with one of our clients. The client has two machines: a desktop and a laptop, both running W10 and on the same wifi network. The desktop has unresolvable connection drops whereas the laptop is buttery smooth.

**Comment by hacker1221** [comment]
> **Solution**:
> Do not use built-in rfc6455 websocket pings and use custom-designed pings over the websocket channel. Connections are stable now.

any solution on how to use customer designed pings

**Comment by qeternity** [comment]
> > **Solution**:
> > Do not use built-in rfc6455 websocket pings and use custom-designed pings over the websocket channel. Connections are stable now.
> 
> any solution on how to use customer designed pings

Have the client send some sort of ping/keepalive payload every X seconds and have the server acknowledge. For each payload the client sends, set a timer for Y seconds that will close and reconnect. If/when the server acknowledges the ping, cancel the timer.

**Comment by codekiln** [comment]
> Have the client send some sort of ping/keepalive payload every X seconds and have the server acknowledge. For each payload the client sends, set a timer for Y seconds that will close and reconnect. If/when the server acknowledges the ping, cancel the timer.
So to summarize, @qeternity, it sounds like you just subvert daphne's ping mechanism and instead send pings from the client?
